File: split/doendv.pas
procedure doendv;
begin
  baseptr := inputptr;
  inputstack[baseptr] := curinput;
  while (inputstack[baseptr].indexfield <> 2) and (inputstack[baseptr].locfield = 0) and (inputstack[baseptr].statefield = 0) do
    baseptr := baseptr - 1;
  if (inputstack[baseptr].indexfield <> 2) or (inputstack[baseptr].locfield <> 0) or (inputstack[baseptr].statefield <> 0) then
    fatalerror(595);
  if curgroup = 6 then
  begin
    endgraf;
    if fincol then
      finrow;
  end
  else
    offsave;
end;

File: cpp/src/doendv.cpp
#include "doendv.h"
#include "erreur.h"
#include "endgraf.h"
#include "fincol.h"
#include "finrow.h"
#include "offsave.h"

void doendv(Token t, AlignRecordNode* &loop)
{
	baseptr = inputptr;
	inputstack[baseptr] = curinput;
	while (inputstack[baseptr].indexfield != v_template && inputstack[baseptr].locfield == 0 && inputstack[baseptr].statefield == token_list)
		baseptr--;
	if (inputstack[baseptr].indexfield != v_template || inputstack[baseptr].locfield || inputstack[baseptr].statefield != token_list)
		fatalerror("(interwoven alignment preambles are not allowed)");
	if (curgroup == align_group)
	{
		endgraf();
		if (fincol(t, loop))
			finrow(loop);
	}
	else
		offsave(t);
}
